File: 2020/cpp/day1/main.cpp
#include "../utils.h"

int main(int argc, char* argv[]) {

  if (argc != 2) {
    std::cerr << "input file is missing! ABORT.\n";
    exit(1);
  }

  // read input from file
  std::vector<int> entries;
  readInput(std::string(argv[1]), entries);


  // part 1
  for (int i = 0; i < entries.size(); i++) {
    for (int j = i + 1; j < entries.size(); j++) {
      int sum = entries[i] + entries[j];
      if (sum == 2020) {
        int product = entries[i] * entries[j];
        std::cout << "Answer Part 1: " << product << "\n";
      }
    }
  }

  // part 2
  for (int i = 0; i < entries.size(); i++) {
    for (int j = i + 1; j < entries.size(); j++) {
      int sum_ij = entries[i] + entries[j];
      if (sum_ij >= 2020) continue;
      for (int k = j + 1; k < entries.size(); k++) {
        int sum = sum_ij + entries[k];
        if (sum == 2020) {
          int product = entries[i] * entries[j] * entries[k];
          std::cout << "Answer Part 2: " << product << "\n";
        }
      } // k
    } // j
  } // i

} 

File: 2020/rust/day1/main.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader, Error, ErrorKind, Read};

fn readInput<R: Read>(io: R) -> Result<Vec<i32>, Error> {
  let br = BufReader::new(io);
  br.lines()
    .map(|line| line.and_then(|v| v.parse().map_err(|e| Error::new(ErrorKind::InvalidData, e))))
    .collect()
}

fn main() -> Result<(), Error> {

  let args: Vec<String> = env::args().collect();

  if args.len() != 2 { panic!("Input file is missing! ABORT.\n"); }

  // read input from file
  let filename = &args[1];
  let entries = readInput(File::open(filename)?)?;

  // part 1
  for i in 0..=entries.len()-1 {
    for j in i..=entries.len()-1 {
      let sum = entries[i] + entries[j];
      if sum == 2020 {
        let product = entries[i] * entries[j];
        println!("Answer Part 1: {}", product);
      }
    }
  }

  // part 2
  for i in 0..=entries.len()-1 {
    for j in i..=entries.len()-1 {
      let sum_ij = entries[i] + entries[j];
      if sum_ij >= 2020 { continue }
      for k in j..=entries.len()-1 {
        let sum = sum_ij + entries[k];
        if sum == 2020 {
          let product = entries[i] * entries[j] * entries[k];
          println!("Answer Part 2: {}", product);
        }
      }
    }
  }

  Ok(())
} 
